File: generate-og.js
#!/usr/bin/env node
/**
 * generate-og.js
 * Generates Open Graph share images for all articles.
 * Run automatically via: npm run build
 * Run manually via: node generate-og.js
 *
 * Reads article front matter, generates a 1200x630 PNG for each
 * into /static/og/<articleId>.png. Skips articles that already
 * have an up-to-date image unless --force is passed.
 */

const fs = require('fs');
const path = require('path');
const { createCanvas } = require('canvas');

const ARTICLES_DIR = path.join(__dirname, 'src/articles');
const OUTPUT_DIR = path.join(__dirname, 'static/og');
const FORCE = process.argv.includes('--force');

// Colors matching site CSS variables
const BG = '#0e0e0e';
const GOLD = '#c9a96e';
const TEXT = '#f0ece4';
const MUTED = '#777777';
const BORDER = '#c9a96e';

function wrapText(ctx, text, maxWidth) {
  const words = text.split(' ');
  const lines = [];
  let line = '';
  for (const word of words) {
    const test = line ? line + ' ' + word : word;
    if (ctx.measureText(test).width > maxWidth && line) {
      lines.push(line);
      line = word;
    } else {
      line = test;
    }
  }
  if (line) lines.push(line);
  return lines;
}

function parseFrontMatter(content) {
  const match = content.match(/^---\n([\s\S]*?)\n---/);
  if (!match) return {};
  const fm = {};
  for (const line of match[1].split('\n')) {
    const colon = line.indexOf(':');
    if (colon === -1) continue;
    const key = line.slice(0, colon).trim();
    const val = line.slice(colon + 1).trim().replace(/^["']|["']$/g, '');
    fm[key] = val;
  }
  return fm;
}

function generateImage(articleId, title, description) {
  const W = 1200, H = 630;
  const canvas = createCanvas(W, H);
  const ctx = canvas.getContext('2d');

  // Background
  ctx.fillStyle = BG;
  ctx.fillRect(0, 0, W, H);

  // Gold left accent line
  ctx.fillStyle = GOLD;
  ctx.fillRect(80, 80, 3, 470);

  // Site label
  ctx.fillStyle = GOLD;
  ctx.font = '500 20px serif';
  ctx.fillText('PARTICIPATORY MIND', 120, 110);

  // Title
  ctx.fillStyle = TEXT;
  ctx.font = 'italic 58px serif';
  const titleLines = wrapText(ctx, title, 900);
  let y = 175;
  for (const line of titleLines.slice(0, 3)) {
    ctx.fillText(line, 120, y);
    y += 72;
  }

  // Description
  ctx.fillStyle = MUTED;
  ctx.font = '26px serif';
  const descLines = wrapText(ctx, description, 900);
  y += 16;
  for (const line of descLines.slice(0, 3)) {
    ctx.fillText(line, 120, y);
    y += 40;
  }

  // Attribution
  ctx.fillStyle = '#444444';
  ctx.font = '20px serif';
  ctx.fillText('Tyler Parker & Claude Sonnet 4.6  —  participatorymind.org', 120, 530);

  // Gold bottom line
  ctx.fillStyle = GOLD;
  ctx.fillRect(80, 578, 1040, 2);

  return canvas.toBuffer('image/png');
}

// Ensure output dir exists
fs.mkdirSync(OUTPUT_DIR, { recursive: true });

// Process all article files
const files = fs.readdirSync(ARTICLES_DIR).filter(f => f.endsWith('.njk'));
let generated = 0, skipped = 0;

for (const file of files) {
  const content = fs.readFileSync(path.join(ARTICLES_DIR, file), 'utf8');
  const fm = parseFrontMatter(content);
  const { articleId, title, description } = fm;

  if (!articleId || !title) { skipped++; continue; }

  const outPath = path.join(OUTPUT_DIR, `${articleId}.png`);

  if (!FORCE && fs.existsSync(outPath)) { skipped++; continue; }

  try {
    const buf = generateImage(articleId, title, description || '');
    fs.writeFileSync(outPath, buf);
    console.log(`  ✓ og/${articleId}.png`);
    generated++;
  } catch (e) {
    console.error(`  ✗ ${articleId}: ${e.message}`);
  }
}

console.log(`\nOG images: ${generated} generated, ${skipped} skipped. Pass --force to regenerate all.`);

File: generate-og.py
#!/usr/bin/env python3
"""
generate-og.py
Generates Open Graph share images for all articles.
Called automatically by: npm run build
Called manually by: python3 generate-og.py
Called with --force to regenerate existing images.

Reads article front matter from src/articles/*.njk
Outputs 1200x630 PNGs to static/og/<articleId>.png
Skips articles that already have an image unless --force is passed.
"""

import os, sys, re, textwrap
from pathlib import Path

try:
    from PIL import Image, ImageDraw, ImageFont
except ImportError:
    print("PIL not found. Run: pip install Pillow --break-system-packages")
    sys.exit(1)

ARTICLES_DIR = Path(__file__).parent / 'src' / 'articles'
OUTPUT_DIR   = Path(__file__).parent / 'static' / 'og'
FORCE        = '--force' in sys.argv

BG    = '#0e0e0e'
GOLD  = '#c9a96e'
TEXT  = '#f0ece4'
MUTED = '#777777'

FONT_PATHS = [
    '/usr/share/fonts/truetype/liberation/LiberationSerif-Italic.ttf',
    '/usr/share/fonts/truetype/dejavu/DejaVuSerif.ttf',
    '/usr/share/fonts/truetype/freefont/FreeSerif.ttf',
]
FONT_PATHS_REG = [p.replace('Italic', 'Regular').replace('-Italic','') for p in FONT_PATHS]

def load_font(paths, size):
    for p in paths:
        if os.path.exists(p):
            return ImageFont.truetype(p, size)
    return ImageFont.load_default()

def parse_front_matter(text):
    m = re.match(r'^---\n(.*?)\n---', text, re.DOTALL)
    if not m: return {}
    fm = {}
    for line in m.group(1).splitlines():
        if ':' not in line: continue
        k, _, v = line.partition(':')
        fm[k.strip()] = v.strip().strip('"\'')
    return fm

def hex_to_rgb(h):
    h = h.lstrip('#')
    return tuple(int(h[i:i+2], 16) for i in (0, 2, 4))

def generate(article_id, title, description):
    W, H = 1200, 630
    img  = Image.new('RGB', (W, H), hex_to_rgb(BG))
    draw = ImageDraw.Draw(img)

    font_title = load_font(FONT_PATHS, 58)
    font_sub   = load_font(FONT_PATHS_REG, 26)
    font_label = load_font(FONT_PATHS_REG, 20)

    # Gold left accent
    draw.rectangle([(80, 80), (83, 550)], fill=hex_to_rgb(GOLD))

    # Site label
    draw.text((120, 92), 'PARTICIPATORY MIND', font=font_label, fill=hex_to_rgb(GOLD))

    # Title
    lines = textwrap.wrap(title, width=26)
    y = 155
    for line in lines[:3]:
        draw.text((120, y), line, font=font_title, fill=hex_to_rgb(TEXT))
        y += 72

    # Description
    y += 16
    desc_lines = textwrap.wrap(description, width=60)
    for line in desc_lines[:3]:
        draw.text((120, y), line, font=font_sub, fill=hex_to_rgb(MUTED))
        y += 40

    # Attribution
    draw.text((120, 526), 'Tyler Parker & Claude Sonnet 4.6  —  participatorymind.org',
              font=font_label, fill=(68, 68, 68))

    # Gold bottom line
    draw.rectangle([(80, 578), (1120, 580)], fill=hex_to_rgb(GOLD))

    return img

OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

files = sorted(ARTICLES_DIR.glob('*.njk'))
generated = skipped = errors = 0

for f in files:
    fm = parse_front_matter(f.read_text())
    article_id  = fm.get('articleId', '')
    title       = fm.get('title', '')
    description = fm.get('description', '')

    if not article_id or not title:
        skipped += 1
        continue

    out = OUTPUT_DIR / f'{article_id}.png'
    if not FORCE and out.exists():
        skipped += 1
        continue

    try:
        img = generate(article_id, title, description)
        img.save(str(out), 'PNG', optimize=True)
        print(f'  ✓  og/{article_id}.png')
        generated += 1
    except Exception as e:
        print(f'  ✗  {article_id}: {e}')
        errors += 1

print(f'\nOG images: {generated} generated, {skipped} skipped, {errors} errors.')
print('Pass --force to regenerate all existing images.')
